### PR TITLE
docs: Add best practices for service discovery implementations

### PR DESCRIPTION
**Pull request by karthi-palanisamy**
This commit addresses the TODO comment in discovery/README.md by adding
comprehensive best practices for implementing service discovery mechanisms.

The new section covers:

- Error handling and retry strategies with exponential backoff
- Performance optimizations (blocking queries, caching, filtering)
- Reliability patterns for handling backend unavailability
- Resource management and cleanup
- Testing requirements
- Code quality guidelines

These practices are derived from existing SD implementations (Consul,
Kubernetes, EC2) and will help contributors create robust, performant service discovery
integrations.

#### Which issue(s) does the PR fix:

N/A - Addresses TODO comment in existing documentation

#### Does this PR introduce a user-facing change?

```release-notes
NONE

### CODE REVIEW

**Comment by bboreham** [comment]
Hello from the bug-scrub!

Thanks for trying to fill in the "TODO".  

I would like to hear from folks who work on Service Discovery - @rexagod or @sysadmind, do you think this is a decent addition?

File: discovery/README.md [hunk 1]
```diff
@@ -263,7 +263,69 @@ Here are some non-obvious parts of adding service discoveries that need to be ve
...
   `<alertmanager_config>` in `docs/configuration/configuration.md`.
 
-<!-- TODO: Add best-practices -->
+### Best Practices for Service Discovery Implementations
+
+When implementing a new service discovery mechanism, follow these best practices to ensure reliability, performance, and maintainability:
+
+#### Error Handling and Retries
+
+- **Implement exponential backoff**: Use exponential backoff with jitter for retry logic to avoid overwhelming the SD backend during outages. See `consul/consul.go` for a reference implementation with `retryInterval` of 15 seconds.
```
Comment (sysadmind) [inline discovery/README.md:272]: The code in `consul/consul.go` doesn't use exponential backoffs that I can see. It's just a simple delay. If we want to recommend exponential backoffs, there should also be reasonable maximums. Something that this guidance doesn't consider is how this affects Prometheus waiting on the results.

File: discovery/README.md [hunk 2]
```diff
@@ -263,7 +263,69 @@ Here are some non-obvious parts of adding service discoveries that need to be ve
...
+  
+- **Handle transient failures gracefully**: Distinguish between temporary network issues and permanent configuration errors. Only retry on transient failures.
+
+- **Set appropriate timeouts**: Configure reasonable timeouts for API calls (e.g., `watchTimeout` of 2 minutes in Consul SD) to prevent blocking indefinitely.
```
Comment (sysadmind) [inline discovery/README.md:276]: I think actual appropriate timeouts would be good to list. Again, the Prometheus runtime likely cares a lot about how long these refreshes take.

File: discovery/README.md [hunk 3]
```diff
@@ -263,7 +263,69 @@ Here are some non-obvious parts of adding service discoveries that need to be ve
...
+
+- **Log errors with context**: Include relevant details (URL, operation, error type) in error logs to aid debugging.
+
+#### Performance Considerations
+
+- **Use blocking queries or long polling**: Where supported by the SD backend (e.g., Consul's blocking queries), use them to reduce unnecessary API calls and improve efficiency.
+
+- **Implement caching**: Cache SD responses where appropriate to reduce load on the SD backend, especially for large deployments.
```
Comment (sysadmind) [inline discovery/README.md:284]: I feel like "where appropriate" is very vague guidance for a best practices. I don't think it's giving much guidance. In my opinion, a best practices doc that is essentially "use your best judgement" is unhelpful.

File: discovery/README.md [hunk 4]
```diff
@@ -263,7 +263,69 @@ Here are some non-obvious parts of adding service discoveries that need to be ve
...
+
+- **Monitor SD health**: Expose metrics via the provided `prometheus.Registerer` to track discovery performance, error rates, and target counts.
+
+#### Resource Management
+
+- **Clean up resources**: Properly close HTTP clients, watchers, and other resources when the SD is stopped. Implement context cancellation handling.
+
+- **Avoid goroutine leaks**: Ensure all goroutines are properly terminated when the discovery Run method's context is cancelled.
+
+- **Limit memory usage**: For SDs that may discover very large numbers of targets, consider streaming or pagination to avoid loading everything into memory at once.
```
Comment (sysadmind) [inline discovery/README.md:308]: This is just regular coding guidance. I don't know that it's useful here without more specific guidance.

File: discovery/README.md [hunk 5]
```diff
@@ -263,7 +263,69 @@ Here are some non-obvious parts of adding service discoveries that need to be ve
...
+
+#### Testing
+
+- **Write comprehensive unit tests**: Cover configuration unmarshalling, metadata extraction, error handling, and edge cases.
+
+- **Test with real data**: Use realistic SD responses in tests to catch unexpected data formats or edge cases.
+
+- **Test failure scenarios**: Verify correct behavior when the SD backend is unavailable, returns errors, or provides malformed data.
+
+- **Validate target group generation**: Ensure target groups are correctly formatted with proper `Source` identifiers and that empty target lists are sent when all targets disappear.
```
Comment (sysadmind) [inline discovery/README.md:318]: An example to  point to would be useful here.

File: discovery/README.md [hunk 6]
```diff
@@ -263,7 +263,69 @@ Here are some non-obvious parts of adding service discoveries that need to be ve
...
+
+#### Code Quality
+
+- **Keep business logic in relabelling**: The SD should extract raw metadata; leave filtering and transformation to relabelling configs. Don't make assumptions about how users will use the discovered targets.
+
+- **Follow existing patterns**: Study similar SDs (e.g., Consul, Kubernetes) for established patterns around configuration, target group management, and error handling.
```
Comment (sysadmind) [inline discovery/README.md:324]: This is essentially saying "do your homework" which again I don't think is that helpful. If there's a really well written SD, that would be good to note as a reference. If there are several patterns, it would be good to call those out.

**Review by sysadmind — COMMENTED** [review]
I think that best practices and more guidance would be useful for SD maintainers. As written, I think that the value of this guidance is marginal. I think it would be much more useful with specific guidance and not such generalized items.
